Supported option to specify resource management for pods in Keycloak CR

mabartos opened the pull request:
Closes #26456 

- When no `limit/requests` set for `memory` in the Keycloak CR and KeycloakRealmImport CR - **set defaults**[1] 
- When no `limit/requests` set for `cpu` in the CRs - **no defaults** (should be managed by the k8s env)
- Different JVM memory settings for Keycloak **running in a container** 
- Set env var `KC_RUN_IN_CONTAINER` flag to mark it's running in a container (set in Dockerfile)
- When Keycloak is running in a container, the memory setting **is relative to the present node**
- Otherwise, the default values are unchanged (`-Xms64m`, `-Xmx512m`)

## Default values

When Keycloak is running in a container, the JVM options `-Xmx`, and `-Xms` are replaced by the JVM options `-XX:MaxRAMPercentage`[2][3][4], and `-XX:InitialRAMPercentage`[2][3][4]. They allow setting relative values to the available resources of a node, thus good candidates for container env.

- `-XX:MaxRAMPercentage` - Maximum heap size as a percentage of total memory
- `-XX:InitialRAMPercentage` - Initial heap size as a percentage of total memory
- `-XX:MinRAMPercentage` (optional) - Maximum heap size as a percentage of total memory (for JVM with a small amount of memory < ~200MB) - default value is 50%, but Keycloak should not be run with avail. memory < ~200MB.

### Analysis
I made an analysis of the memory utilization of Keycloak in order to provide some reasonable default values. As we know, the total memory consists of several elements, so it needs to be taken into the equation. With the `-Xmx` and `-XX:MaxRAMPercentage`, we're touching only the heap.

I tried to set the `requests` memory for containers in the operator to some middle value - not having high requirements for potential node candidates, not having so much committed memory, and executing a reasonable count of GC executions. But the `requests` define the worst-case scenario of resources Keycloak would work with, in some reasonable manner - ratio performance/needed-resources.

We can consider the overall memory as heap + non-heap. 
_NOTE: Values described below are gathered as approximations of several runs of one Keycloak instance with Podman and with the default values described in the next section._

The overall non-heap memory during my analysis (even with gathering profiling info) was approximately **~200MB**, consisting of Metaspace (~130MB), Code cache(~50MB), Compressed class space (~20MB) + some minor additionals. 

During the startup, the heap is utilized quite much, as there are several requirements for the object allocations + GC management + other stuff. But after stabilizing and performing GC executions, the RSS is ~70MB. 

### Suggested default values

**Operator**:
`resources.requests.memory` - 768 MiB - Requested memory
`resources.limits.memory` - 4 GiB - Maximum memory

**JVM**:
`-XX:MaxRAMPercentage` - 70% of total requested memory
`-XX:MinRAMPercentage` - 70% of total requested memory
`-XX:InitialRAMPercentage` - 50% of total requested memory
 
![resourcesKeycloak-final](https://github.com/keycloak/keycloak/assets/38039883/e22aa4d1-1673-4596-81ae-5c5421a8df41)

I tried to be close with the heap size values to current values in the Keycloak distribution (`-Xmx512m`). Thus, in the worst case, the maximum heap size will be ~538MiB (70% of the 768MiB). The initial heap size will be bigger than the current one (`-Xms64m`), in order to prevent multiple GC executions during the startup, so I chose ~384MiB (50% of the 768MiB).

You can see that 230MiB is reserved for the non-heap section, which IMHO is sufficient for the worst-case scenario (as during the analysis, max. 200MiB was used by JVM). 

[1] https://github.com/keycloak/keycloak/pull/26661/files#diff-f53c77df31edb884b75c19bd62de359748d617a94909dbafece4bbc435c0ee23R12-R14
[2] https://developers.redhat.com/articles/2022/04/19/java-17-whats-new-openjdks-container-awareness#tuning_defaults_for_containers
[3] https://www.baeldung.com/java-jvm-parameters-rampercentage
[4] https://www.ibm.com/docs/en/sdk-java-technology/8?topic=options-xxinitialrampercentage

mabartos commented:
I'll summarize the resource analysis to be clear why these values were chosen and then, the PR will be ready to review.

mabartos commented on quarkus/dist/src/main/content/bin/kc.sh:
```diff
@@ -96,7 +96,21 @@ if [ -z "$JAVA_OPTS" ]; then
+
+   if [ -z "$JAVA_OPTS_KC_HEAP" ]; then
+      JAVA_OPTS_KC_HEAP="-Xms64m -Xmx512m"
+      if [ "$KC_RUN_IN_CONTAINER" = "true" ]; then
```
Unfortunately, I haven't found any way how to determine that it's running in a container. I tried to leverage JDK's container awareness, but unsuccessfully.

vmuzikar reviewed (commented):
@mabartos Thanks for the changes, nice work.

In general LGTM, however I'd like to resolve the following points before we merge:
1. Instead of hardcoding the defaults, I'd suggest making them a config option. (See comment below.)
2. `KC_RUN_IN_CONTAINER` env var should not be set by the Operator as it AFAIK is not specific to it.
3. I'd like to ask @ahus1 for a review mainly from the default values perspective.

mabartos commented on operator/src/main/java/org/keycloak/operator/Constants.java:
```diff
@@ -71,4 +72,8 @@ public final class Constants {
     public static final String KEYCLOAK_HTTP_RELATIVE_PATH_KEY = "http-relative-path";
+
+    // Container requests/limits resources (tested suitable minimum for well-performant Keycloak)
+    public static final Quantity KEYCLOAK_RESOURCES_MEMORY_REQUESTS_DEFAULT = new Quantity("768Mi");
+    public static final Quantity KEYCLOAK_RESOURCES_MEMORY_LIMITS_DEFAULT = new Quantity("4Gi");
```
@vmuzikar I've been thinking about it, and I'm not sure if that's necessary. In the initial version of this PR, I put the capability to change these defaults, but then, I realized that it might not be necessary at all. Do we need to provide other config options for this purpose? 

To be more specific, when the user wants to set different values, the `limits/requests` should be used for this purpose. Why should different properties be set for it when the Operator sets it all for these managed containers? 

@vmuzikar Or do you see any benefit for doing it?

vmuzikar replied:
Yes, the defaults are not meant to be changed by the user. But IMHO we should still have it configurable as an escape hatch (if for whatever reason the defaults need to be different for that particular Operator installation). This would also align with other default values like Keycloak image coordinates.

mabartos replied:
@vmuzikar ok, works for me

vmuzikar commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:
```diff
@@ -54,6 +54,11 @@
+    /**
+     * ENV var for notifying distribution that it is running inside a container
+     */
+    public static final String KC_RUN_IN_CONTAINER = "KC_RUN_IN_CONTAINER";
```
> but what about custom Dockerfiles

I might be wrong but isn't the [`ENV` instruction](https://docs.docker.com/engine/reference/builder/#env) inherited by any child images? It should be present in any custom images as long as they use our image as the base.

mabartos replied:
Yep, you're right! :)) Didn't come to my mind.

mabartos commented on operator/src/main/resources/application.properties:
```diff
@@ -9,6 +9,9 @@ kc.operator.keycloak.image=${RELATED_IMAGE_KEYCLOAK:quay.io/keycloak/keycloak:ni
 kc.operator.keycloak.image-pull-policy=Always
 kc.operator.keycloak.start-optimized=false
 kc.operator.keycloak.poll-interval-seconds=60
+# Keycloak container default requests/limits resources
+kc.operator.keycloak.resources.memory-requests=768Mi
+kc.operator.keycloak.resources.memory-limits=4Gi
```
@vmuzikar Does this format work for you?

mabartos replied:
@shawkins I was considering that approach, but then I chose the less verbose in terms of the Keycloak config interface (SmallRye). But I have no preference here, we can do it that way to line up with the kube approach.

mabartos replied:
@shawkins @vmuzikar Ok, I'll amend it.

vmuzikar commented on operator/src/main/java/org/keycloak/operator/crds/v2alpha1/realmimport/KeycloakRealmImportSpec.java:
```diff
@@ -29,6 +31,10 @@ public class KeycloakRealmImportSpec {
     private RealmRepresentation realm;
 
+    @JsonProperty("resources")
+    @JsonPropertyDescription("Compute Resources required by Keycloak container")
```
I'd add something like:
```suggestion
    @JsonPropertyDescription("Compute Resources required by Keycloak container; if left unspecified, the value is inherited from Keycloak CR")
```

vmuzikar reviewed (commented):
Couple of more thoughts.

* We should probably also add some note to the [Advanced Operator guide](https://github.com/keycloak/keycloak/blob/750bc2c09c63a81aba02ebb6574b2b69fc6386de/docs/guides/operator/advanced-configuration.adoc) that scaling memory is also reflected in the JVM.
* We need to add some tests for Realm Import.
* It'd be nice to also add some ne validation for resources to [`validatePodTemplate`](https://github.com/keycloak/keycloak/blob/750bc2c09c63a81aba02ebb6574b2b69fc6386de/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java#L212).

mabartos commented:
@vmuzikar @shawkins I addressed your comments and put the explanation of chosen default values to the description of this PR (cc: @ahus1)

---------------------

@vmuzikar @shawkins Do you know how I can inject the Operator config to the ImportRealm dependent resource (`KeycloakRealmImportJobDependentResource`)? I tried to inject it via CDI `@Inject` annotation, but NPE is thrown when accessing. See here[1]. Any help appreciated :)) 

[1] https://github.com/keycloak/keycloak/actions/runs/7873658615/job/21481902905?pr=26661#step:8:3694

shawkins commented:
@mabartos due to prior limiations in the operator sdk we needed to construct the KeycloakRealmImportJobDependentResource instance in the KeycloakRealmImportController as part of the manual workflow control. The underlying issues may have been fixed in this release if you want to try changing the KeycloakRealmImportJobDependentResource to be injected into the KeycloakRealmImportController.

https://github.com/quarkiverse/quarkus-operator-sdk/issues/685
https://github.com/operator-framework/java-operator-sdk/issues/2010

vmuzikar reviewed (commented):
@mabartos Thank you for the changes and the elaborate justification of the defaults. Very nice write-up!

From my perspective LGTM, except from the points I mentioned in my [previous review](https://github.com/keycloak/keycloak/pull/26661#pullrequestreview-1875469067).

mabartos committed:
Supported option to specify resource management for pods in Keycloak CR

Closes #26456

Co-authored-by: Václav Muzikář <vaclav@muzikari.cz>
Signed-off-by: Martin Bartoš <mabartos@redhat.com>

mabartos commented on operator/src/main/java/org/keycloak/operator/Utils.java:
```diff
@@ -77,4 +80,35 @@ public static <T extends HasMetadata> Optional<T> getByName(Class<T> clazz, Func
+        if (requestsMemory != null) {
+            var specifiedMemoryIsLessThanDefault = requestsMemory.getNumericalAmount().intValue() < defaultRequestsMemory.getNumericalAmount().intValue();
+            if (specifiedMemoryIsLessThanDefault) {
+                Log.warnf("Provided 'requests' memory ('%s') is less than used default value ('%s'). Use it in your risk, as Keycloak performance might be degraded.", requestsMemory, defaultRequestsMemory);
```
jfyi; I've added a warning message about the requested memory being less than the default one.

shawkins replied:
@mabartos this warning will happen on every reconciliation, so it will need to be at a debug level and/or a status warning instead.

mabartos replied:
ok, thanks

mabartos commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java:
```diff
@@ -234,6 +234,9 @@ public void validatePodTemplate(Keycloak keycloakCR, KeycloakStatusAggregator st
                                 "The image of the keycloak container cannot be modified using podTemplate");
                     }
+                    if (container.getResources() != null) {
+                        status.addWarningMessage("Resources requirements of the Keycloak container cannot be modified using podTemplate");
```
@vmuzikar Not sure, what validation should be done here? The resources should not be specified via podTemplate, when used as first citizen. So, is it sufficient?

mabartos reviewed (commented):
@vmuzikar @shawkins @Pepo48 @ahus1 Should be ready for review.

mabartos commented on operator/src/main/java/org/keycloak/operator/Utils.java:
```diff
@@ -77,4 +80,35 @@ public static <T extends HasMetadata> Optional<T> getByName(Class<T> clazz, Func
+        if (requestsMemory != null) {
+            var specifiedMemoryIsLessThanDefault = requestsMemory.getNumericalAmount().intValue() < defaultRequestsMemory.getNumericalAmount().intValue();
+            if (specifiedMemoryIsLessThanDefault) {
+                Log.debugf("Provided 'requests' memory ('%s') is less than used default value ('%s'). Use it in your risk, as Keycloak performance might be degraded.", requestsMemory, defaultRequestsMemory);
```
@shawkins Is it ok, or do you suggest to use the `KeycloakStatusAggregator` ?

shawkins replied:
Yes, it's just awkward to use currently because it's associated with the top level reconciliation of the keycloak cr, not any of the dependents.

mabartos commented on docs/documentation/release_notes/topics/24_0_0.adoc:
```diff
@@ -195,6 +195,32 @@ spec:
+----
+
+For more details, check the
+https://www.keycloak.org/operator/advanced-configuration[Operator Advanced configuration].
```
I'm not sure if there's a way to specify even the section(newly created) and have the ExternalLinksTest pass :) 

Keeping it as is for now. 

We should also reorder the new things for KC 24 doc, as all changes to Keycloak CR should be somewhere together (follow-up issue).

vmuzikar commented on operator/src/test/java/org/keycloak/operator/testsuite/integration/RealmImportTest.java:
Thanks for extending the tests! We now test the defaults as well as that the Keycloak CR can be used to configure resources on the Job. I think we're missing a test for actually setting resources on the Realm Import CR.

vmuzikar reviewed (commented):
@mabartos Thank you for the changes! Added a last few comments.

I'd like to also ask @ahus1 for a review of the chosen defaults.

mabartos commented on docs/documentation/release_notes/topics/24_0_0.adoc:
```diff
@@ -195,6 +195,32 @@ spec:
       key: config.xml
 ----
 
+= Keycloak CR resources options
```
Ok

vmuzikar replied:
I think this is still valid. IMHO it'd be good to mention the new default requests and limits in the upgrading guide.

mabartos commented on docs/documentation/release_notes/topics/24_0_0.adoc:
```diff
@@ -195,6 +196,38 @@ spec:
+      memory: 3Gi
+----
+
+When the `resources` options are not specified, these default values are used:
+
+* `requests.memory` = 768Mi
+* `limits.memory` = 4Gi
+
+For more details, check the
+https://www.keycloak.org/operator/advanced-configuration[Operator Advanced configuration].
```
@vmuzikar Is it sufficient? I'd say so, as it points to the more comprehensive guide.

mabartos commented on operator/src/test/java/org/keycloak/operator/testsuite/integration/RealmImportTest.java:
```diff
@@ -142,14 +155,23 @@ public void testWorkingRealmImport() {
         deployKeycloak(k8sclient, keycloak, false);
 
+        final var resourceRequirements = new ResourceRequirements();
+        resourceRequirements.setLimits(Map.of(
+                "memory", new Quantity("3", "G")));
+        resourceRequirements.setRequests(Map.of(
+                "memory", new Quantity("600", "M")));
+
         // Act
-        K8sUtils.set(k8sclient, getClass().getResourceAsStream("/example-realm.yaml"));
+        KeycloakRealmImport realmImport = K8sUtils.getResourceFromFile("/example-realm.yaml", KeycloakRealmImport.class);
+        realmImport.getSpec().setResourceRequirements(resourceRequirements);
+        K8sUtils.set(k8sclient, realmImport);
```
@vmuzikar jfyi - Set the resources requirements for KeycloakRealmImport CR

ahus1 commented on docs/guides/server/containers.adoc:
```diff
@@ -229,4 +229,26 @@ podman|docker run --name keycloak_unoptimized -p 8080:8080 \
+The {project_name} container, instead of specifying hardcoded values for the initial and maximum heap size, uses relative values to the total memory of a node.
+This behavior is achieved by JVM options `-XX:MaxRAMPercentage=70`, and `-XX:InitialRAMPercentage=50`.
+
+The `-XX:MaxRAMPercentage` option represents the maximum heap size as 70% of total node memory.
+The `-XX:InitialRAMPercentage` option represents the initial heap size as 50% of the total node memory.
```
This mentions "node", where it should probably mention "container"? The node's memory is not relevant to the JVM IMHO.

mabartos replied:
@ahus1 You're right, thanks

mabartos committed:
Amendments for docs

Signed-off-by: Martin Bartoš <mabartos@redhat.com>

ahus1 commented:
*TL;DR* IMHO the memory defaults described above can cause some problems. It might be simpler to set them to the same value as default or rather close to each other. We can have a separate issue for that if we want to get this one merged. 

**The longer version:**

I read in the description above, that the JVM percentages are about the requested memory. IMHO this is not what is happening, as IMHO the JVM checks the memory limit and uses this for percentages. Otherwise the JVM would never use more memory than requested, which would be a strange thing.

I see that the default settings for memory deviate quite much between requested and limit. This is what happens when they deviate: 

* Kubernetes allocates the Pod to a node which has (based on all "requested" memory limits for all the pods already on that node) enough memory for the "requested" memory limit of that Pod. 
* Once the JVM then uses more than the requested limit (and other might do as well), the node will eventually get into MemoryPressure state, and Pods are evicted from that node and restarted on other nodes
* If the JVM uses more memory in a spike, the Kubelet might not register the the MemoryPressure in time, and instead Linux will OOM one of the processes on the node (not necessarily Keycloak). See https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#kubelet-may-not-observe-memory-pressure-right-away for an explanation. 

The more the requested deviated from the limit, the more likely it is that the "strange" things described above will happen.

At the same time, if we set the the MaxRAMPercentage to 70%, the natural setting for the requested is to be 70% of the limit, as we expect the JVM to only use 70% of the memory limit.

So after discussing in the Cross-DC I'd suggest to choose either 1 GB or 2 GB as the limit: 

* Set it to 1000 MB limit /  700 MB requested if we want to show that Keycloak is small for those who just want to try it
* Set it to 2000 MB limit / 1400 MB requested if we want to allow a medium sized production deployment

When using those memory settings, it might be good to add `-XX:+PrintFlagsFinal` to see what results the calculation based on the percentage came up with.

ahus1 commented:
Update on the 70% calculation in the previous comment: The KCB will need to run some tests to see how much the JVM limits itself to that 70%, or if it uses more memory than that when for example the number of threads increases. Again, this is beyond this issue.

ahus1 reviewed (dismissed):
Approving from Cross DC point of view. We would be good use the PR as is, use the KCB project to run experiments, and then report back on different defaults.

vmuzikar reviewed (commented):
@mabartos Thank you for the update!

@shawkins Do you want to review as well?

mabartos committed:
Add upgrading guide

Signed-off-by: Martin Bartoš <mabartos@redhat.com>

mabartos commented:
@vmuzikar I've added the upgrading guide. I hope it's sufficient.

vmuzikar reviewed (approved):
@mabartos Thank you, LGTM!

@shawkins Feel free to merge if you want to review and looks good to you too.

mabartos commented:
> I read in the description above, that the JVM percentages are about the requested memory. IMHO this is not what is happening, as IMHO the JVM checks the memory limit and uses this for percentages. Otherwise the JVM would never use more memory than requested, which would be a strange thing.

@ahus1 Probably, I wasn't so clear about the description above. It was meant to reflect the worst-case scenario when on a node, there's only the minimal (requested) available memory, so it's also the limit (requests=limits). _I'm not considering k8s scheduling algorithm or stuff around it_.  It's supposed to be present that even in the worst-case, Keycloak should have the necessary memory for non-heap memory + the same max heap as for the Keycloak distribution (~500M).

> So after discussing in the Cross-DC I'd suggest to choose either 1 GB or 2 GB as the limit

Yep, I'm not against it, and thanks for the explanation of the possible situations; makes sense.

@ahus1 I'd be really glad if you could run some experiments and reflect some real-world scenarios on these default values. Thanks!

mabartos commented:
@ahus1 @shawkins So, are you ok to merge it now and then amend these defaults in a follow-up task?

ahus1 commented:
@mabartos - I'm ok to merge (see also the comments in the chat). Defaults will be handled in a separate issue.

shawkins reviewed (approved):
LGTM, thank you @mabartos fantastic work

ahus1 commented:
Intermediate test results in KCB: I've seen in smaller tests the JVM report non-heap memory usage of 240 MB, so for the calculation we'll use 300 MB assuming this is on the safe side. 

The current attempt to document this for users will be: 
1. Use the sizing guide to determine the requested memory (for example 1250 MB)
2. To determine the memory limit, substract the expected non-heap memory usage (300 MB), and then divide the remaining result by 0.7 (the 70% used in Keycloak). This would then result in a memory limit of ~1360 MB.

We'll need to run longer tests, and also chaos tests, as I expect load  spikes with lots of active threads to increase the non-heap memory usage. Let's see.

vmuzikar commented:
@ahus1 Thanks for testing this change! If I understand correctly, from this initial testing we can't still assume any changes to the default values, correct? We'll see based on further tests.

ahus1 commented:
@vmuzikar - the latest tests showed up to 277 MB committed nonheap memory. Due to that, I assume that 300 MB non-heap memory is on the safe side. 

The sizing guide has been updated in https://github.com/keycloak/keycloak/issues/27078. I leave it for the CND team to decide if they want to update any documentation.

vmuzikar commented:
@ahus1 Thanks for getting back with the results. So the suggestion is to change the defaults for request to 1 GB and keep limit as is?

ahus1 commented:
@vmuzikar - reading the docs, the default limit is 4GiB.

Using the formula above, the JVM would use at startup 50% of that as the heap, and from the experiments 300 MB as non-heap. At runtime, the JVM would use up to 70% of the total memory as heap, again plus up to 300 MB non-heap memory. 

This would result in a memory usage of between 2.3GiB and 3.1GiB of memory. So if you want to keep the limit of 4GiB, I would recommend setting the requested memory to 3.1GiB to ensure Kubernetes allocates the Pod to a node that has that much memory available from a scheduling perspective. 

But doing this would make Keycloak look quite fat compared to previous releases. 

If you want to go for 1 GiB requested memory, and using the formula above, you would need to set a limit of 1 GiB. This is actually also the lowest limit for this setup, as if you would go below the limit of 1 GiB limit, the 70% heap memory taking from this don't leave enough space for the 300 MB non-heap memory.

ahus1 commented:
@vmuzikar and I discussed the following: 

* A medium sized production deployment would have a memory limit of 2 GiB
* The requested memory should need to be as close as possible to the actual memory usage of the JVM, as the requested resources are used for scheduling the Pod on a node. https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#how-pods-with-resource-requests-are-scheduled
* Given the formula above, the JVM will see the limit of 2GiB, and set 70% of that as the max heap which would be 1.4GiB. In addition to that, the non-heap usage of the JVM will be 0.3GiB. This sums up as an expected memory usage of 1.7GiB 
* The expected memory usage of 1.7 GiB would then also be the requested limit to ensure the node has at least this memory available for Keycloak
* Should the non-heap usage exceed the 0.3GiB, the JVM can use the gap between 1.7GiB and 2.0GiB as long as there is memory available on the node. 

@vmuzikar agreed to create a follow-up issue.

vmuzikar commented:
@ahus1 Thank you for the summary! Created https://github.com/keycloak/keycloak/issues/27169.
